File: 1331-path-with-maximum-gold/path-with-maximum-gold.java
class Solution {
    public int [] X = new int[]{1,-1,0,0};
    public int [] Y = new int[]{0,0,1,-1};
    public int [][] visited;
    public int getMaximumGold(int[][] grid) {
        this.visited = new int[grid.length][grid[0].length];
        int result = 0;
        for(int i = 0;i<grid.length;i++){
            for(int j = 0;j<grid[0].length;j++){
                if(grid[i][j] > 0){
                    result = Math.max(result,fun(i,j,grid));
                }
            }
        }
        return result;
    }
    public int fun(int x, int y,int [][] grid){
        visited[x][y] = 1;
        int val = 0;
        for(int i = 0;i<4;i++){
            int nextX = x + X[i];
            int nextY = y + Y[i];
            if(nextX >= 0 && nextX < grid.length && nextY >= 0 && nextY < grid[0].length && this.visited[nextX][nextY] != 1 && grid[nextX][nextY] > 0){
                val = Math.max(val,fun(nextX,nextY,grid));
            }
        }
        visited[x][y] = 0;
        return grid[x][y] + val;
    }
}

File: 1331-path-with-maximum-gold/path-with-maximum-gold.py
"""

1,0,7
2,0,6
3,4,5
0,3,0
9,0,20


"""
class Solution:
    def getMaximumGold(self, grid):
        rows, cols = len(grid), len(grid[0])
        dirs = [(1,0),(0,1),(-1,0),(0,-1)]
        def dfs(i, j):
            gold = grid[i][j]
            grid[i][j] = 0
            max_gold = 0
            for di, dj in dirs:
                ni, nj = i + di, j + dj
                if 0 <= ni < rows and 0 <= nj < cols and grid[ni][nj] > 0:
                    max_gold = max(max_gold, dfs(ni, nj))
            grid[i][j] = gold
            return gold + max_gold
        result = 0
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] > 0:
                    result = max(result, dfs(i, j))
        return result

        